Allow regrading of solved problems

dmartin opened the pull request:
Previous behavior (did not actually verify submitted flag if the team had already solved the problem):

| previously submitted? | correct? | message displayed |
| --- | --- | --- |
| N | N | (error) "That is incorrect!" |
| N | Y | (success) "That is correct!" |
| Y | N | (error) "Flag correct: however, you have already received points for that flag." |
| Y | Y | (error) "Flag correct: however, you have already received points for that flag." |

New behavior:

| previously submitted? | correct? | message displayed |
| --- | --- | --- |
| N | N | (error) "That is incorrect!" |
| N | Y | (success) "That is correct!" |
| Y | N | (error) "Flag incorrect: please note that you or your team have  already solved this problem." |
| Y | Y | (success) "Flag correct: however, you have already received points for that flag." |

#215 notes that we may want to consider storing submissions after a team's first correct submission in the database. I've (tentatively) decided not to do that because it seems like it could make the correct/incorrect submissions breakdown in the admin view less meaningful, though this could easily be changed.

resolves #215

hiliang-cmu commented:
Getting into particulars of #215 flag storage, there would definitely be use in having multiple 'correct' submissions with differing `uid`s, especially with a platform view in mind. 

So perhaps the current solely `tid`-based approach of problem solve states can be expanded to cover `uid` as well: 

When an incorrect flag is submitted on a problem solved by a different team member, store as usual **until** the current user solves it himself.

dmartin commented:
Modified such that:
>When an incorrect flag is submitted on a problem solved by a different team member, store as usual until the current user solves it himself.

and added different messages depending on whether the user themselves, or another team member, has previously solved a problem.

Will open a separate issue to discuss potential storing of additional valid submissions.
